File: packages/aitrack-lib/src/readers/cursor/jwt.ts
import { Buffer } from 'node:buffer';

/**
 * Reading the subject out of Cursor's access token.
 *
 * The payload is untrusted remote input, so every step here has to tolerate
 * garbage rather than assume a well-formed token.
 */
export function decodeJwtPayload(token: string): { sub?: string } | null {
  const encodedPayload = token.split('.', 2)[1];
  if (!encodedPayload) return null;
  const base64 = encodedPayload.replaceAll('-', '+').replaceAll('_', '/');
  const padded = base64.padEnd(Math.ceil(base64.length / 4) * 4, '=');
  try {
    return JSON.parse(Buffer.from(padded, 'base64').toString('utf8')) as { sub?: string };
  } catch {
    return null;
  }
}

File: crates/aitrack-core/src/readers/cursor/jwt.rs
//! JWT payload decoding for Cursor access tokens.

use base64::prelude::*;
use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
pub struct JwtPayload {
    pub sub: Option<String>,
}

/// Decode JWT payload (untrusted remote input, tolerates garbage).
pub fn decode_jwt_payload(token: &str) -> Option<JwtPayload> {
    let parts: Vec<&str> = token.split('.').collect();
    let encoded_payload = parts.get(1)?;

    // Convert URL-safe base64 to standard base64
    let base64 = encoded_payload.replace('-', "+").replace('_', "/");

    // Add padding
    let padded = format!("{}{}", base64, "=".repeat((4 - base64.len() % 4) % 4));

    // Decode base64
    let decoded = BASE64_STANDARD.decode(padded.as_bytes()).ok()?;
    let json_str = String::from_utf8(decoded).ok()?;

    // Parse JSON
    serde_json::from_str(&json_str).ok()
}
